ASoC: SOF: Intel: don't check number of sdw links when set dmic_fixup

bardliao opened the pull request:
Currently, we assume that the PCH DMIC pins are pin-muxed with SoundWire links. However, we do see a HW design that use PCH DMIC along with 3 SoundWire links. Remove the check now.
With this change BIOS should only report the presence of PCH DMI when it is truly exist.

bardliao committed:
ASoC: SOF: Intel: don't check number of sdw links when set dmic_fixup

Currently, we assume that the PCH DMIC pins are pin-muxed with SoundWire
links. However, we do see a HW design that use PCH DMIC along with 3
SoundWire links. Remove the check now.
With this change the PCM DMIC will be presented if it is reported by the
BIOS irrespective of whether there are SDW links present or not.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao commented:
@ranj063 Do you see any risk that we merge this PR?

ranj063 commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1321,22 +1321,8 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
-			if (mach->tplg_quirk_mask & SND_SOC_ACPI_TPLG_INTEL_DMIC_NUMBER)
-				dmic_fixup = true;
-		}
+		if (sdw_mach_found || mach->tplg_quirk_mask & SND_SOC_ACPI_TPLG_INTEL_DMIC_NUMBER)
```
@bardliao can we please fix the typo in commit message "PCH DMI". Also, the last line somehow reads as though the BIOS is going to report differently because of this change. But rather what you're trying to say is that we check for PCM DMIC reported by the BIOS irrespective of whether there are SDW links present or not.

ranj063 replied:
and also in the case where the SDW and DMIC are pin-muxed, will this change have a detrimental effect on machine driver probe because we ended up choosing the wrong topology name?

bardliao replied:
Yes, if the BIOS report PCH DMIC but there is no PCH DMIC, we will choose the wrong topology name. That's why I am not confident with this change. But on the other hand, that is the BIOS's responsibility to report the real HW configuration and kernel should trust the BIOS.

ranj063 replied:
I think we don't have a choice but to trust the bios @ujfalusi @lgirdwood what do you think?

bardliao replied:
I vaguely remember we did this for RVPs. Not sure if it happens in real products. The worst case is that someone reports an issue that no sound card is created, and we may need to fix it with some quirk.

ujfalusi replied:
@bardliao, you are too optimistic ;) We have seen broken BIOS information and I'm sure we are continue to see them.

But imho we still need to trust the information BIOS provides, it is most likely correct..

bardliao replied:
@ranj063 @ujfalusi Alternative solution is to add `SOC_SDW_PCH_DMIC` quirk to this specific device. What do you think?

lgirdwood replied:
I would make any assumption obvious in the logs i.e. if we BIOS says we have DMIC & SDW then we can log this with a ```if this is wrong pls set "force_fix_something=1" in module command line```.
This way it should be obvious to a user on how to remedy if BIOS is reporting incorrect data.

bardliao replied:
Log added

lgirdwood commented on sound/soc/sof/intel/hda.c:
```diff
@@ -527,6 +527,8 @@ static int check_dmic_num(struct snd_sof_dev *sdev)
 		dmic_num = intel_nhlt_get_dmic_geo(sdev->dev, nhlt);
 
 	dev_info(sdev->dev, "DMICs detected in NHLT tables: %d\n", dmic_num);
+	dev_info(sdev->dev,
+		 "If the DMIC number doesn't match to the real HW design, please overwrite it with \"dmic_num\" kernel param.\n");
```
I would change this too
```c
dev_info(sdev->dev, "DMICs detected in NHLT tables: %d, can be overidden with \"dmic_num\" kernel param if incorrect\n", dmic_num);
```
and I would move the new larger warning to the code block when we detect both SDW and DMIC
```c
dev_info(sdev->dev, "NHLT reporting both SDW and DMIC endpoints, if incorrect, pls override with .....);
```

bardliao replied:
Update: only show the warning when both SDW mic and PCH DMIC are present.

bardliao committed:
ASoC: Intel: sof_sdw: warn both sdw and pch dmic are used

Typically, SoundWire MIC and PCH DMIC will not coexist. However, we may
want to use both of them in some special cases. Add a warning to let
users know that SoundWire MIC and PCH DMIC are both present and they
could overwrite it with kernel params.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ranj063 commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1321,22 +1321,8 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
-			if (mach->tplg_quirk_mask & SND_SOC_ACPI_TPLG_INTEL_DMIC_NUMBER)
-				dmic_fixup = true;
-		}
+		if (sdw_mach_found || mach->tplg_quirk_mask & SND_SOC_ACPI_TPLG_INTEL_DMIC_NUMBER)
```
@bardliao is the check for sdw_mach_found important? I think we only add the dmic prefix to the tplg name if either the tplg quirk is set or the num_dmics reported is non-zero isnt it?

bardliao replied:
Yes, sdw_mach_found is important because we don't add -2ch/-4ch postfix for SSP machines.

lgirdwood commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -1180,6 +1182,18 @@ static int sof_card_dai_links_create(struct snd_soc_card *card)
+		for (i = 0; i < num_ends; i++) {
+			if (sof_end->dai_info->dai_type == SOC_SDW_DAI_TYPE_MIC) {
+				dev_warn(dev,
+					 "Both SDW DMIC and PCH DMIC are present, if incorrect, please override with kernel params\n");
```
which params ? lets be more helpful.

lgirdwood reviewed (commented):
LGTM, just need a string update

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1321,22 +1321,8 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
-			 * link 2 and 3, or link 1 and 2, thus we only try to enable dmics
-			 * if all conditions are true:
-			 * a) 2 or fewer links are used by SoundWire
-			 * b) the NHLT table reports the presence of microphones
```
maybe you only need to refine these criteria?

You can have 3 SoundWire links along 1 DMIC link. But I don't think you can have 3 SoundWire links with 2 DMIC links - since in the latter case you need at least 3 wires, 4 if the clock isn't shared.

bardliao replied:
@plbossart Hmm, the question is that do DMIC links always share with SoundWire links and will the SoundWire link number increase? Is it possible that someday we have 3 SoundWire links and 2 DMIC links?

bardliao commented:
@ujfalusi @ranj063 Are you good with the PR?

ujfalusi commented on sound/soc/intel/boards/sof_sdw.c:
```diff
@@ -919,6 +921,11 @@ static int create_sdw_dailink(struct snd_soc_card *card,
+			if (sof_end->dai_info->dai_type == SOC_SDW_DAI_TYPE_MIC &&
+			    mach_params->dmic_num > 0) {
+				dev_warn(dev,
+					 "Both SDW DMIC and PCH DMIC are present, if incorrect, please set kernel params snd_sof_intel_hda_generic dmic_num=0 to disable PCH DMIC\n");
```
and what if the PCH-DMIC is the correct and the SDW DMIC is not?

bardliao replied:
Not all SDW DMIC dai are implemented as optional and can be removed. We add `.quirk = SOC_SDW_CODEC_MIC,` in the codec_info_list[] if the SDW DMIC DAI is optional. And can remove it with snd-soc-sof-sdw quirk. But we need to add the `SOC_SDW_CODEC_MIC` quirk if the `.quirk_exclude` flag is set in the codec_info_list[] or remove the `SOC_SDW_CODEC_MIC` quirk if the `.quirk_exclude` flag is not set to remove the SDW DMIC. That's why I didn't add description for removing the SDW DMIC here.
